abort/readme.md

```diff
@@ -0,0 +1,220 @@
+3. `signal` ([_`Event`_]) — `Event` that will trigger abort of effect calls, accepts key of effect call to be aborted. If multiple effect calls share the same key, they all will be aborted on signal triggered with that key
+4. `domain` ([_`Domain`_]) - optional effector domain. If provided - all internal units will be created from it
+
+### Returns
+
+- `result` ([_`Effect`_]) — Abortable Effect. If aborted manually by provided signal, will return `AbortedError` in `effect.failData` event. `AbortedError` can be detected by `instanceof AbortedError` or by `aborted` and `key` fields in it
+
+[_`event`_]: https://effector.dev/docs/api/effector/event
+[_`effect`_]: https://effector.dev/docs/api/effector/effect
+
+### Examples
+
+#### Abort all effect calls
+
+```ts
+import { createEvent } from 'effector';
+import { abort } from 'patronum/abort';
+
+const cancel = createEvent<string>();
+const abortableFx = abort({
+  signal: cancel,
+  getKey: () => 'shared',
+  async handler(_: void, { onAbort }) {
+    await new Promise((r, rj) => {
+      let id = setTimeout(r, 1000);
+
+      onAbort(() => {
+        // cancel promise and cleanup timeout
+        rj();
+        clearTimeout(id);
+      });
+    });
+  },
+});
+
+abortableFx();
+abortableFx();
+abortableFx();
+
+cancel('shared');
+
+// calling cancel with "shared" key will abort all abortableFx calls in fly with AbortedError
+```
+
+#### Abort specific effect call by key
+
+```ts
+import { createEvent, sample } from 'effector';
+import { abort } from 'patronum/abort';
+
+const cancel = createEvent<number>();
+const abortableFx = abort({
+  signal: cancel,
+  getKey: (param: number) => param,
+  async handler(params: number, { onAbort }) {
+    await new Promise((r, rj) => {
+      let id = setTimeout(r, params);
+
+      const revoke = onAbort(() => {
+        console.log('bla');
+        // you can call onAbort multiple times and revoke reigstered callbacks
+      })
+
+      revoke() // removes specific abort callback
+
+      onAbort(() => {
+        // cancel promise and cleanup timeout
+        rj();
+        clearTimeout(id);
+      });
+    });
+  },
+});
+
+const stop = createEvent();
+
+sample({
+  clock: stop,
+  fn: () => 1000,
+  target: cancel,
+});
+
+abortableFx(300);
+abortableFx(1000);
+abortableFx(10000);
+
+stop();
+
+// calling cancel with "1000" key will abort only specific effect call with "1000" key
+```
+
+#### Check if aborted
+
+```ts
+import { createEvent, guard } from 'effector';
+import { abort, AbortedError } from 'patronum/abort';
+
+const cancel = createEvent<string>();
+const abortableFx = abort({
+  signal: cancel,
+  getKey: () => 'shared',
+  async handler(_: void, { onAbort }) {
+    await new Promise((r, rj) => {
+      let id = setTimeout(r, 1000);
+
+      onAbort(() => {
+        // cleanup
+        rj();
+        clearTimeout(id);
+      });
+    });
+  },
+});
+
+guard({
```

Comment:
Maybe it's better to return from `abort` an object `{ abortableFx, abortedEvent }` 🤔

Reply:
I'm not sure about that 🤔

User still would have to manually filter all `fail` and `failData` calls 🤷

Reply:
Maybe something like this would do?

```ts
const { 
 fx,
 aborted, // filtered fail, only with AbortedError
 failed // filtered fail, only if not AbortedError
} = abort({ ... })
```

Reply:
Yeah, I like this idea.

Reply:
This looks weird IMO.
What about wrappers?

```javascript
import { abort, aborted, failed, abortedData, failedData } from "patronum/abort"

const  fx = abort({ ... })

const store = createStore()
  .on(failed(fx.fail), () => { ... })
  .on(aborted(fx.fail), () => { ... })
  .on(failedData(fx.failData), () => { ... })
  .on(abortedData(fx.failData), () => { ... })
```

Reply:
It's kind of verbose 🤔 

Maybe something like this?

```ts
const fx = abort({ ... })

const store = createStore(null)
  .on(failed(fx), () => { ... })
  .on(aborted(fx), () => { ... })
  .on(failedData(fx), () => { ... })
  .on(abortedData(fx), () => { ... })
```

Reply:
Maybe it could be both :)

```javascript
const aborted = (unit: Event | Effect) => {
  const event = is.effect(unit) ? unit.fail : unit
  return event.filterMap(
    payload =>
      payload.error instanceof AbortError
        ? payload
        : undefined
  )
}
```

Reply:
@zerobias what do you think?

Considering that in the future the effector may also get the abortable effects feature at the core (correspondingly, migration from the patronum variant should be as simple as possible), which option is better?

Reply:
@AlexandrHoroshih separate methods which may be applied to arbitrary effects going to be an ambiguous approach and ability of the cancellation to pass through nested effects still does not seem obvious. If you remove a requirement to derive abort triggers from existed effects and support for nesting then api for that triggers may fit configuration object

```ts
const timeoutFx = abort({
  on: {
    abort: [timeoutAborted],
    done: [timeoutCompleted],
    fail: [setTimeoutNotSupported]
  },
  handler(params, {onAbort}) {
    
  }
})
```

And real-world case with timeouts rise another important question: how we would derive id from params if `TimeoutID` could be given only during execution at first place? Artificial ids are not an option. In other words, how application should communicate with body of the effect?

Reply:
You need your implementation to pass reality-check, even a link to repl with code which _really_ do something would be enough to found such an issue